subsys/bluetooth/services/cgms/cgms.c

```diff
@@ -0,0 +1,414 @@
+ */
+int bt_cgms_add_measurement(uint16_t glucose)
+{
+	/*Prepare the measurement */
+	struct bt_cgm_meas_t meas;
+
+	meas.flag = 0;
+	meas.glucose_concentration = glucose;
+	meas.time_offset = (k_uptime_get_32() - local_start_time) / 1000 / 60;
+	meas.sensor_status_annunciation.warning = 0;
+	meas.sensor_status_annunciation.calib_temp = 0;
+	meas.sensor_status_annunciation.status = 0;
+
+	memcpy(&latest_meas, &meas, sizeof(meas));
+
+	int rc;
+	/*Add the measurement into DB */
+	rc = cgms_racp_add_meas(meas);
+
+	return rc;
+}
+
+void report_meas(struct k_work *item)
+{
+	/* If it is zero, we should not notify the client.
+	 * But we still reschedule the task with minimum
+	 * value of comm_interval to monitor it is enabled
+	 * in the future.
+	 */
+	if (comm_interval > 0) {
+		/* Notify all connected peers when it is periodic notification.*/
+		bt_conn_foreach(BT_CONN_TYPE_LE, bt_cgms_notify_meas, &latest_meas);
+	}
+	k_work_reschedule(&report_meas_work, K_MINUTES(
+		comm_interval > 0 ? comm_interval : BT_CGMS_COMM_INTERVAL_MIN));
+}
+
+int cgms_init(void)
+{
+	int rc;
+
+	memset(&latest_meas, 0, sizeof(latest_meas));
+
+	comm_interval = BT_CGMS_COMM_INTERVAL_MIN;
+	k_work_init_delayable(&report_meas_work, report_meas);
+	rc = k_work_reschedule(&report_meas_work, K_MINUTES(comm_interval));
+
+#ifdef CONFIG_BT_CGMS_LOC_FINGER
+	feature.sample_location = BT_SERVICE_CGMS_MEAS_LOC_FINGER;
+#endif
+#ifdef CONFIG_BT_CGMS_LOC_AST
+	feature.sample_location = BT_SERVICE_CGMS_MEAS_LOC_AST;
+#endif
+#ifdef CONFIG_BT_CGMS_LOC_EAR
+	feature.sample_location = BT_SERVICE_CGMS_MEAS_LOC_EAR;
+#endif
+#ifdef CONFIG_BT_CGMS_LOC_CONTROL
+	feature.sample_location = BT_SERVICE_CGMS_MEAS_LOC_CONTROL;
+#endif
+#ifdef CONFIG_BT_CGMS_LOC_SUB_TISSUE
+	feature.sample_location = BT_SERVICE_CGMS_MEAS_LOC_SUB_TISSUE;
+#endif
+#ifdef CONFIG_BT_CGMS_LOC_NOT_AVAIL
+	feature.sample_location = BT_SERVICE_CGMS_MEAS_LOC_NOT_AVAIL;
+#endif
+
+#ifdef CONFIG_BT_CGMS_TYPE_CAP_BLOOD
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_CAP_BLOOD;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_CAP_PLASMA
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_CAP_PLASMA;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_VEN_BLOOD
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_VEN_BLOOD;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_VEN_PLASMA
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_VEN_PLASMA;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_ART_BLOOD
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_ART_BLOOD;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_ART_PLASMA
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_ART_PLASMA;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_UNDET_BLOOD
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_UNDET_BLOOD;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_UNDET_PLASMA
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_UNDET_PLASMA;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_FLUID
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_FLUID;
+#endif
+#ifdef CONFIG_BT_CGMS_TYPE_CONTROL
+	feature.type = BT_SERVICE_CGMS_MEAS_TYPE_CONTROL;
+#endif
+	feature.feature = 0;
+
+	status.status.status = 0;
```

Comment:
The status charcteristic has two fields, time offset and status. I will pop up time offset. For status, as it can only has non-zero value when optional feature is supported. As this implementaion only has mandatory features, the status field will remains zero.